File: src/macroGenerator.js
import OpenAI from 'openai';
import fs from 'fs/promises';

export class MacroGenerator {
  constructor(apiKey) {
    this.openai = new OpenAI({ apiKey });
  }

  async readFullHTML(htmlPath) {
    const htmlContent = await fs.readFile(htmlPath, 'utf-8');
    return htmlContent;
  }

  async generateMacroCode(task, htmlPath, previousAttempts, selectedModel) {
    const htmlContent = await this.readFullHTML(htmlPath);

    let prompt = `
You are a web automation expert. Generate Playwright JavaScript code to perform the following task:

Task: ${task.description}

HTML Content:
${htmlContent}

Generate clean, working Playwright code that:
1. Uses import syntax (not require) since this is ES module
2. Opens the HTML file using fileUrl parameter
3. Performs the requested action
4. Includes proper waits and error handling
5. Takes screenshots using screenshotsDir parameter with path.join()
6. Returns detailed result object with action summary and data extracted

IMPORTANT:
- Use ES6 import syntax, NOT require()
- Use the provided parameters: page, fileUrl, screenshotsDir
- Import path if needed: import path from 'path'
- Return detailed object with extracted data

Return only the JavaScript code without any markdown formatting or explanation.
`;

    // Add error context from previous attempts for multi-turn improvement
    if (previousAttempts && previousAttempts.length > 0) {
      const errorContext = previousAttempts
        .filter(attempt => !attempt.success && attempt.error)
        .map(attempt => `Attempt ${attempt.attemptNumber}: ${attempt.error}`)
        .join('\n');

      if (errorContext) {
        prompt += `

PREVIOUS ERRORS TO AVOID:
${errorContext}

Please fix these issues in your new code.`;
      }
    }

    const response = await this.openai.chat.completions.create({
      model: selectedModel,
      messages: [{ role: 'user', content: prompt }],
      temperature: 0.3,
    });

    return response.choices[0].message.content;
  }
}

File: python_src/macro_generator.py
import os
import json
from openai import OpenAI
from bs4 import BeautifulSoup
from pathlib import Path

class MacroGenerator:
    def __init__(self, api_key):
        self.client = OpenAI(api_key=api_key)
    
    def analyze_html(self, html_path):
        """HTML 파일을 분석하여 구조 정보 추출"""
        with open(html_path, 'r', encoding='utf-8') as f:
            html_content = f.read()
        
        soup = BeautifulSoup(html_content, 'lxml')
        
        structure = {
            'forms': [],
            'buttons': [],
            'inputs': [],
            'links': []
        }
        
        # Forms 분석
        for form in soup.find_all('form'):
            structure['forms'].append({
                'id': form.get('id'),
                'action': form.get('action'),
                'method': form.get('method')
            })
        
        # Buttons 분석
        for button in soup.find_all(['button', 'input']):
            if button.name == 'input' and button.get('type') != 'submit':
                continue
            structure['buttons'].append({
                'text': button.text.strip() if button.text else button.get('value'),
                'id': button.get('id'),
                'class': ' '.join(button.get('class', [])),
                'type': button.get('type')
            })
        
        # Inputs 분석
        for input_elem in soup.find_all('input'):
            if input_elem.get('type') == 'submit':
                continue
            structure['inputs'].append({
                'name': input_elem.get('name'),
                'id': input_elem.get('id'),
                'type': input_elem.get('type'),
                'placeholder': input_elem.get('placeholder')
            })
        
        # Links 분석
        for link in soup.find_all('a'):
            structure['links'].append({
                'text': link.text.strip(),
                'href': link.get('href'),
                'id': link.get('id')
            })
        
        return html_content, structure
    
    def generate_macro_code(self, instruction, html_path):
        """LLM을 사용하여 매크로 코드 생성"""
        html_content, structure = self.analyze_html(html_path)
        
        prompt = f"""
You are a web automation expert. Generate Python Playwright code to perform the following task:

Task: {instruction}

Page Structure:
{json.dumps(structure, indent=2)}

HTML Content (first 2000 chars):
{html_content[:2000]}

Generate clean, working Python Playwright code that:
1. Opens the HTML file
2. Performs the requested action
3. Includes proper waits and error handling
4. Takes screenshots to document the action
5. Returns the result of the action

Return only the Python code without any markdown formatting or explanation.
The code should be a function that takes page, file_url, and screenshots_dir as parameters.
"""
        
        response = self.client.chat.completions.create(
            model='gpt-4-turbo-preview',
            messages=[{'role': 'user', 'content': prompt}],
            temperature=0.3,
        )
        
        return response.choices[0].message.content